### ISSUE TITLE
[QUERY] Upload / download of stream-backed content to Azure Blob Storage with limited memory footprint

### ISSUE THREAD

**Issue by white-sagittarius** [Azure/azure-sdk-for-java#18295]
**Query/Question**
I was not successful getting streaming file upload (file 1.17GB) with limited heap (512MB) to work using azure-storage-blob:12.10.0-beta.1 (also tested on 12.9.0).

Posted a question to StackOverflow on the matter: https://stackoverflow.com/questions/65395726/upload-of-large-files-using-azure-sdk-for-java-with-limited-heap

We need to upload / download files to / from Azure Blob Storage in non-memory demanding way (service heap is limited with `-Xmx512m`).

Both `BlobClient.upload` and `BlockBlobClient.upload` use a lot of heap presumably loading all file content into memory. Using direct copy to `BlobOutputStream` results in file corruption.

```
BlobContainerClient blobContainerClient = new BlobContainerClientBuilder()
  .connectionString("****")
  .containerName("****")
  .buildClient();

long blockSize = DataSize.ofMegabytes(4L).toBytes();

BlockBlobClient blockBlobClient = blobContainerClient.getBlobClient("image.jpg").getBlockBlobClient();

// create new blob / erase content of existing blob
blockBlobClient.commitBlockList(List.of(), true);

BlockBlobOutputStreamOptions options = (new BlockBlobOutputStreamOptions()).setParallelTransferOptions(
  (new ParallelTransferOptions()).setBlockSizeLong(blockSize).setMaxConcurrency(1).setMaxSingleUploadSizeLong(blockSize));

File image = new File("/Users/user/Downloads/Airbus_Pleiades_50cm_8bit_RGB_Yogyakarta.jpg");

try (InputStream is = new FileInputStream(image)) {
  try (OutputStream os = blockBlobClient.getBlobOutputStream(options)) {
    IOUtils.copy(is, os); // copies input to output using 8KB buffer
  }
}

log.info("EXPECTED SIZE: {}; ACTUAL SIZE: {}", image.length(), blockBlobClient.getProperties().getBlobSize());
``` 

Code snipped above completes successfully and outputs:
```
18:00:24.577 [main] INFO my.play.quicktest.misc.CheckAzureStreamingUpload - EXPECTED SIZE: 41026764; ACTUAL SIZE: 41026764
```

However, uploaded image file contains artifacts (only top 1/3 of image is preserved, while other 2/3 are all grey pixels - at least in Azure Portal Edit window and after manual download).

**Question:** how do I get a working upload / download solution with small memory footprint using azure-sdk-for-java?

***Why is this not a Bug or a feature Request?***
The documentation for stream-based upload / download with azure-storage-blob is scarce, not particularly sure I am doing the right thing.

**Setup (please complete the following information if applicable):**
 - OS: macOS Catalina 10.15.7
 - IDE : IntelliJ IDEA 2020.3
 - 12.10.0-beta.1 and 12.9.0
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by rickle-msft** [comment]
@white-sagittarius Thank you for opening this issue. Let me spend a little time trying to reproduce this and get back to you. Our first order of business needs to be ensuring we aren't corrupting your data, and if we are to fix that

**Comment by rickle-msft** [comment]
@white-sagittarius While I'm investigating this, did you use the same configurations for the BlobClient.upload method? In terms of block size and concurrency?

**Comment by white-sagittarius** [comment]
@rickle-msft I used the signature that does not specify upload options (i.e. block size & concurrency):
```
blobClient.upload(is, image.length(), true /* overwrite */);
blockBlobClient.upload(is, image.length(), true /* overwrite */);
```

Do you want me to check `uploadWithResponse` method with the same `BlobParallelUploadOptions`?

My goal is to limit azure-storage-blob memory usage as much as possible while still having a decent upload / download throughput (for instance, using 8KB block size has proven not to be an option as we need hundreds of thousands of requests and we also run out of 50K limit of blob blocks / 100K limit of uncommitted blob blocks for larger files).

I was playing with `maxSingleUploadSize` and `blockSize` options as well as tried to override `reactor.bufferSize.x` / `reactor.bufferSize.small` to limit Reactor queues memory usage (as suggested [here](https://github.com/Azure/azure-sdk-for-java/issues/10823#issuecomment-626082185))

**Comment by white-sagittarius** [comment]
I tried the approach suggested [here](https://stackoverflow.com/a/65403169/14865538) and it worked pretty nice. Memory consumption is fairly low, file is not getting corrupted. The only inconvenience is necessity to have actual file (we are receiving input as `InputStream` and would have to write it to a temporary file for leveraging `BlobClient.uploadFromFile` / `BlobClient.downloadToFileWithResponse` APIs).

![image](https://user-images.githubusercontent.com/76443987/102862430-c393cc80-4439-11eb-8eb2-62528ef47951.png)

I suppose, `azure-storage-blob` uses file directly for upload which enables to skip quite a few copying / buffering steps?

**Comment by white-sagittarius** [comment]
@rickle-msft by the way, I noticed that only 6 files are uploaded / downloaded in parallel. If I work with more than 6 files in parallel, the rest is waiting for the first 6 to complete. Do you happen to know if there is a setting to control this?

**Comment by rickle-msft** [comment]
@white-sagittarius Thanks for all this information. Responding to your items in order:

- Yes, if you could try configuring the ParallelTransferOptions when you call blobClient.upload(), that would also be a good route to investigate. If you limit blockSize and maxConcurrency, then you can more effectively limit the amount of memory we'll use at a given time. The javadocs for those fields should explain that in more depth.
- One thing to note, blockBlobClient.upload() will try to stream the entire source in one go. So if you have any unreliability in your network, that may cause some difficulty. It also won't try doing anything in parallel, and depending on your service version/account configuration, it's possible that the file you're trying to upload will be too big for that method anyway as there's a lower size cap. blobClient.upload() will do some buffering and that allows us to break up the blob and upload it in parallel, which also offers a large size cap on the blob. Just something to keep in mind when experimenting with these methods. 
- You are correct that when the data is in a file we can elide most of the buffering steps. In case it's helpful, the primary reason we have to buffer for non-file inputs at all is to support retries, which we find to be critical in production environments for distributed apps. It also allows us to ensure that we send the full block size at once and aren't unnecessarily limiting the maximum size of the blob
- Can you share more about how you're trying to parallelize the 6+ files? Perhaps a code snippet? Generally parallelization at this level is up to the customers to implement, but depending on what route you're taking I may be able to offer some guidance or insights.

I believe I have also confirmed that there is some corruption happening in BlockBlobOutputStream. I am going to investigate that more today, and I created a tracking issue for that specifically.

**Comment by rickle-msft** [comment]
@white-sagittarius Looking more into the outputStream issue. If it's possible, can you try changing your stream copy logic to use a 10k buffer instead of an 8k buffer? In fiddling around trying different configurations, I'm seeing an 8k buffer always fails on inputs larger than about 10mb, but I'm seeing a 10k buffer always works when all else is constant, which is somewhat mysterious to me

**Comment by rickle-msft** [comment]
@white-sagittarius You also mentioned you see a lot of grey in your image. It's interesting that the corrupted portion is uniform and not just garbage. Do you know if grey corresponds to a particular byte value in your image format? A bunch of 0s perhaps?

**Comment by rickle-msft** [comment]
@white-sagittarius I think I've resolved what's going on with the data corruption. My hypothesis is that there is a buffer being reused in your stream copy code and it's being updated from the source stream faster than the output stream can consume it. This is what I found happening in my repro. I've shared my stream copy code below. Originally it just created a byte[] for the copy once at the start of the operation and then it reused that. Adding the line I commented on below allowed my tests to pass. We actually do perform a copy of this data internally fairly early on, but I think having the stream copy code in a tight loop like this is writing fast enough such that it still happens before our copy.

I imagine that adding extra buffering isn't going to make your memory constraints happy, but if you could verify this fix to get the corruption concerns out of the way, then we can move on to resolving the concerns around memory consumption.

As an aside, although I mentioned we do a copy internally, we might be able to investigate moving that copy further up in the stack to narrow the window for this sort of thing. I think the copy also happens as an async operation, which is why this is possible in the first place rather than the read-copy-send-repeat pattern being strictly sequential.  

```
byte[] retrievedBuff = new byte[copyBufferSize];
        int nextCopy = (int) Math.min(retrievedBuff.length, writeLength);
        int count = sourceStream.read(retrievedBuff, 0, nextCopy);
        long length = 0;

        while (nextCopy > 0 && count != -1) {
            if (outStream != null) {
                outStream.write(retrievedBuff, 0, count);
            }

            length += count;

            retrievedBuff = new byte[copyBufferSize]; // adding this line fixed the corruption issue.
            nextCopy = (int) Math.min(retrievedBuff.length, writeLength - length);
            count = sourceStream.read(retrievedBuff, 0, nextCopy);
        }
```

**Comment by white-sagittarius** [comment]
@rickle-msft you are absolutely right about buffer being reused. We leverage [`IOUtils.copy`](https://github.com/apache/commons-io/blob/master/src/main/java/org/apache/commons/io/IOUtils.java#L863). Internally it uses the same byte array to transfer 8KB chunks from input stream to output stream.

Checked upload with no buffer reuse - no file corruption, everything worked as expected:
```
public static void main(String[] args) throws IOException {
  BlobContainerClient blobContainerClient = new BlobContainerClientBuilder()
    .connectionString("****")
    .containerName("****")
    .buildClient();

  long blockSize = DataSize.ofMegabytes(4L).toBytes();

  BlockBlobClient blockBlobClient = blobContainerClient.getBlobClient("image.jpg").getBlockBlobClient();

  // create new blob / erase content of existing blob
  blockBlobClient.commitBlockList(List.of(), true);

  BlockBlobOutputStreamOptions options = (new BlockBlobOutputStreamOptions()).setParallelTransferOptions(
    (new ParallelTransferOptions()).setBlockSizeLong(blockSize).setMaxConcurrency(1).setMaxSingleUploadSizeLong(blockSize));

  File image = new File("/Users/user/Downloads/Airbus_Pleiades_50cm_8bit_RGB_Yogyakarta.jpg");

  try (InputStream is = new FileInputStream(image)) {
    try (OutputStream os = blockBlobClient.getBlobOutputStream(options)) {
      //IOUtils.copy(is, os); // copies input to output using 8KB buffer
      int n;

      byte[] buffer = new byte[8 * 1024]; // initial 8KB buffer

      while (IOUtils.EOF != (n = is.read(buffer))) {
        os.write(buffer, 0, n);

        buffer = new byte[8 * 1024]; // new 8KB buffer
      }
    }
  }

  log.info("EXPECTED SIZE: {}; ACTUAL SIZE: {}", image.length(), blockBlobClient.getProperties().getBlobSize());
}
```

Please, note, that buffering data while copying between source & destination with buffer reuse is a well-known and widely adopted IO pattern in Java, so it should be a clearly indicated in the documentation if this use case is not supported by `BlobOutputStream`. As far as I can tell, only synchronous copying of the provided array in `write​(byte[] b, int off, int len)` will prevent this issue, due to array `b` potentially changing as soon as method call returns (asynchronous flow without copying is problematic).

My assumption that there is some hidden limitation of up to 6 files uploaded / downloaded concurrently was preposterous as I discovered that it was due to Google Chrome settings - I was using [Swagger](https://swagger.io/) to call upload / download endpoints ([_Chrome has a limit of 6 connections per host name, and a max of 10 connections_](https://blog.bluetriangle.com/blocking-web-performance-villain))  

I plan to move forward with file-based implementation as it has low memory footprint and seemingly no corruption risks. The only downside is necessity to use **store-and-forward** approach, that limits throughput and requires writing to considerably slower medium on the host instead of doing immediate streaming to client (i.e. there are actually 2 uploads / downloads each time: client => service, service => azure / azure => service, service => client)

**Comment by rickle-msft** [comment]
@white-sagittarius Thanks for giving that a try, and glad to hear it helped. I understand that reusing buffers in this way is ideal, so we'll have some discussions on our team after the new year to figure out how to support that scenario. I think this will also be a good time to evaluate the end to end story of all these upload options and optimize the number of copies we do in general. I also agree with your suggestions of synchronous copies and, as a last resort, documentation.

Glad to hear a resolution to the 6 files question as well.

While we're working on/discussing this, it sounds like there are at least some functional work arounds to these issues, even if they aren't yet ideal. We will keep you posted on the result of these conversations and hopefully have some improvements for you soon. Thank you again for your feedback!

**Comment by rickle-msft** [comment]
I recently merged a PR that addresses this issue (#18598), so I am going to close it as I believe we have also discussed all other questions in the thread to resolution. Please let me know if you have further issues or questions.
